### ISSUE TITLE
gpr_mu_lock assertion when trying to make RPCs from iOS

### ISSUE THREAD

**Issue by jcanizales** [grpc/grpc#3364]
Something between commits fb956aa878bddc7a55f2baea760d3c90a7822be5 (worked) and 64b357ce8ffac125c21981dfdb795502562a263a (broken) broke iOS.

Stack trace (I can get line numbers and local variables values if needed):
<img width="1440" alt="screen shot 2015-09-17 at 1 24 44 pm" src="https://cloud.githubusercontent.com/assets/5120183/9945293/3fe70daa-5d41-11e5-8107-8b9cc842ad50.png">

**Comment by jcanizales** [comment]
Commit 705a49a81270794ea04a93a2d6e0afb53f27d0f7 was good too.

**Comment by yang-g** [comment]
I saw this too at 9ed4be48955e3b65b0540f70060b7191d8c1eef1.

**Comment by yang-g** [comment]
stack trace

```
#0  0x00007f019fcd3cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f019fcd70d8 in __GI_abort () at abort.c:89
#2  0x00007f01a09af998 in gpr_mu_lock (mu=0x1eb0760)
    at src/core/support/sync_posix.c:48
#3  0x00007f01a0c29436 in grpc_pollset_add_fd (pollset=0x1eb0758, fd=0x1ead350)
    at src/core/iomgr/pollset_posix.c:143
#4  0x00007f01a0c2a4e3 in grpc_pollset_set_add_fd (pollset_set=0x1eb55e8, 
    fd=0x1ead350) at src/core/iomgr/pollset_set_posix.c:111
#5  0x00007f01a0c2c1b9 in grpc_tcp_client_connect (cb=0x7f01a0c389d1 <connected>, 
    arg=0x7f0198000aa0, interested_parties=0x1eb55e8, addr=0x7f0198001a80, 
    addr_len=28, deadline=...) at src/core/iomgr/tcp_client_posix.c:264
#6  0x00007f01a0c38c34 in connector_connect (con=0x7f0198000aa0, 
    args=0x7f019eee8930, result=0x7f0198001cb8, notify=0x7f0198001cd0)
    at src/core/surface/channel_create.c:103
#7  0x00007f01a0c1c4f4 in grpc_connector_connect (connector=0x7f0198000aa0, 
    in_args=0x7f019eee8930, out_args=0x7f0198001cb8, notify=0x7f0198001cd0)
    at src/core/client_config/connector.c:48
#8  0x00007f01a0c208ca in continue_connect (c=0x7f0198001c70)
    at src/core/client_config/subchannel.c:318
#9  0x00007f01a0c216f2 in on_alarm (arg=0x7f0198001c70, iomgr_success=1)
    at src/core/client_config/subchannel.c:634
#10 0x00007f01a0c25dce in run_some_expired_alarms (drop_mu=0x7f01a1044e80 <g_mu>, 
    now=..., next=0x7f019eee8e60, success=1) at src/core/iomgr/alarm.c:351
#11 0x00007f01a0c25eae in grpc_alarm_check (drop_mu=0x7f01a1044e80 <g_mu>, 
    now=..., next=0x7f019eee8e60) at src/core/iomgr/alarm.c:363
#12 0x00007f01a0c27756 in background_callback_executor (ignored=0x0)
    at src/core/iomgr/iomgr.c:72
#13 0x00007f01a09afda5 in thread_body (v=0x1eb29d0)
    at src/core/support/thd_posix.c:57
#14 0x00007f019f604182 in start_thread (arg=0x7f019eee9700)
    at pthread_create.c:312
#15 0x00007f019fd9747d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

**Comment by jcanizales** [comment]
Mine is only happening for SSL: cleartext works fine.

**Comment by jcanizales** [comment]
The crash doesn't happen if `GRPC_OP_SEND_INITIAL_METADATA` isn't passed to `grpc_call_start_batch`.

**Comment by jcanizales** [comment]
@jboeuf, might this be related to the recent auth plugin work? This client doesn't use it, but "only after send_metadata is called over SSL" sounds pretty close to that area of the code.

**Comment by jboeuf** [comment]
The auth plugin work has not been merged yet to master (#3176). @ctiller has been adding locks to fix a memory leak lately in the security code (see https://github.com/grpc/grpc/commit/131f6edafa1a281604fffc24019dcb99134eec9f). Could that be related?

**Comment by yang-g** [comment]
I now think our issues may not be the same one. The one I had were about
channel connectivity checker timed out but did not remove the pollset
registered with the channel fast enough. Then when a channel state change
comes, it will try to hold the pollset which may be gone with the
completion queue. The issue could have been in existence all the time. Any
way, I might create a new issue.

**Comment by jcanizales** [comment]
I've confirmed that this bug was introduced by PR #3338.

**Comment by jcanizales** [comment]
Specifically, by commit 131f6edafa1a281604fffc24019dcb99134eec9f.
